### ISSUE TITLE
Replace contentlayer.dev with markdowndb

### ISSUE THREAD

**Issue by rufuspollock** [datopian/datahub-next#51]
As per parent #32 we want to replace contentlayer data calls with markdowndb.

## Acceptance

See PR #56 

Following are replaced

- [x] `pages/[[...slug]].jsx 2:import { allDocuments }` **✅ 2023-03-16 [...slug].tsx now uses MarkdownDB to fetch the paths and load the documents**
- [x] `/content/getters/docs.js 1:import { allDocs }` **✅ 2023-03-16 allDocs now uses MarkdownDB to fetch the files inside the docs folder; docs/index.md moved to pages/docs.tsx;**
- [x] `content/getters/blogs.js 1:import { allBlogs }` used for blogs. **✅ 2023-03-16 allBlogs now uses MarkdownDB to fetch the files inside the blog folder; blog/index.md moved to pages/blog.tsx**
- [x] `content/getters/collections.js  1:import { allCollections }` **✅ 2023-03-16 same as for docs and blogs**
- [x] `scripts/search.mjs 2:import { allDocuments } from "../.contentlayer/generated/index.mjs";` **✅2023-03-16 DELETED**
- [x] `lib/getAuthorsDetails.js 2:import { allPeople }` **✅2023-03-16 fixed, it now uses MarkdownDB**
- [x]  `pages/_all.jsx 2:import { allPages }` **✅2023-03-16 fixed, it now uses MarkdownDB and our own render pipeline**

Things to fix for the future (moved to parent issue #32)

## Tasks

- [x] Where is contentlayer.dev API used? **✅2023-03-06 see below**

Fix all the following

- [x] `pages/[[...slug]].jsx 2:import { allDocuments }`
	- [x] first get working on `pages/test/slug` directory then do `pages/...slug` **✅ 2023-03-16 made pages/...test work as part of the development process, pages/...slug is working as expected**
	- [x] How to fix markdown with imports in e.g. index.md? **Fix: easiest is just to convert to js / ts files in pages directory**
	  - What's the problem? we won't be able to have imports in markdown anymore with our new pipeline. This is a problem for index.md and maybe enterprise/index.md
	  - NB: this solution will require one other fix: remove optional aspect of catch-all.
	- [x] move `index.md` to `pages/index.js` and probably enterprise stuff **✅ 2023-03-16 index, enterprise, blog, collections and docs are now tsx files**
	  - [x] change ``[[...slug]].js` => ``[...slug].js` + `index.js` we are fine.
- [x] `/content/getters/docs.js 1:import { allDocs }`  ⏭️ convert this to a normal static page `pages/docs.tsx` and load from markdowndb
  - [x] Used in `content/docs/index.md` to fetch the list of documents, and then this list is passed to `<DocsList />` (the data sections are not hardcoded).
- [x] `content/getters/blogs.js 1:import { allBlogs }` used for blogs.
- [x] `content/getters/collections.js  1:import { allCollections }`
- [x] `scripts/search.mjs 2:import { allDocuments } from "../.contentlayer/generated/index.mjs";` ⏭️ DELETE (we aren't using search atm)
- [x] `lib/getAuthorsDetails.js 2:import { allPeople }`
- [x]  `pages/_all.jsx 2:import { allPages }` ⏭️ easy replace or even just remove _all.jsx

### MarkdownDB fixes

- [x] Should that be a singleton? **Probably, though easy way is to just to check if db file already on disk and return that and if not create it and then return it**
- [x] Can indexFolder also handle the creation of the db file? **✅2023-03-15 yes: have createAndIndex function**

### Where is contentlayer.dev API used

Results of `ag contentlayer/generated`

```
content/getters/docs.js
1:import { allDocs } from "contentlayer/generated";

content/getters/blogs.js
1:import { allBlogs } from "contentlayer/generated";

content/getters/collections.js
1:import { allCollections } from "contentlayer/generated";

scripts/search.mjs
2:import { allDocuments } from "../.contentlayer/generated/index.mjs";

lib/getAuthorsDetails.js
2:import { allPeople } from "contentlayer/generated";

pages/_all.jsx
2:import { allPages } from "contentlayer/generated";

pages/[[...slug]].jsx
2:import { allDocuments } from "contentlayer/generated";
```

**Comment by demenech** [comment]
**Updates:** 

will review the work and draft a PR tomorrow. So far:

- Had to transform markdowndb into a package so that it can be built and imported in `next.config.mjs` in order to run `indexFolder` only once. Will check if there's any other way of doing this.
  -  Reason is that we can't import `.ts` files in `.mjs` files, and `next.config.mjs` cannot be changed to a ts file because Next.js does not support that.
  - The result is the following folder structure:
    ![ss-2023-03-14-18:17:26](https://user-images.githubusercontent.com/16550934/225138428-560fad12-4e2c-4703-9e72-a51f71b6e901.png)
  - Added a `build` command to `lib/markdowndb/package.json`
- Removed search.mjs
- Replaced
  - allPeople
  - allCollections
  - allBlogs

### Important note:

I think replacing allDocuments is out of the scope of this ticket, because the `body.code` property of the objects of this array is used to render the pages at https://github.com/datopian/datahub-next/blob/main/pages/%5B%5B...slug%5D%5D.tsx#L32.

**Comment by demenech** [comment]
**Updates:** 

- Draft PR: https://github.com/datopian/datahub-next/pull/56
  - Code from the previous reviewed and updated.
- I spent some time trying to understand how contentlayer.dev builds its `next-contentlayer` module, because I wanted to make the `indexFolder` import use this syntax  `import { withContentlayer } from "next-contentlayer";`. Created similar tsconfig and package.json files and tried building markdowndb as CommonJs, but no success so far. (not that important for now)
- Currently working on making test/[...slug].tsx work, still have to create index.tsx and replace imports

**Comment by rufuspollock** [comment]
👍👏

> I spent some time trying to understand how contentlayer.dev builds its next-contentlayer module, because I wanted to make the indexFolder import use this syntax import { withContentlayer } from "next-contentlayer";. Created similar tsconfig and package.json files and tried building markdowndb as CommonJs, but no success so far. (not that important for now)

This seems a nice to have so definitely leave for now.

Let's focus on getting something working and then iterate.

**Comment by rufuspollock** [comment]
Notes from chat today.

- [x] How should we handle layouts? **✅2023-03-16 we have to switch in rendering steps on layout type as we did in Flowershow**
  - [x]x Will we have "layouts" ie. markdown file can specify a different page layout? **✅2023-03-16 yes**
  - Where is it set? **✅2023-03-16 it has to be set as a e.g. `layout` attribute on frontmatter (or added later)**
  - How do i handle this temporarily where we don't have a mechanism for computed fields in markdowndb loading? **✅2023-03-16 Currently manually setting layout to "blog" for files in the blog folder so that the pages look the same**
- [x] `_relative_path` is a to `_url_path` because it's much more convenient for the replacement for now.
- [x] Re blogs, collections and index are being transformed into .tsx files. Do we have a plan to handle markdown pages with imported components in the future? Or is that something we are just not going to support? **✅2023-03-16 we can't support it b/c it requires files on local disk and we don't want that. More specifically: next-mdx-remote does not support it. For that we need mdx-bundler *and* files on local disk**
  - To be clear:
    - We can support idea that different markdown files "types" get access to different sets of components
    - What we are *not* supporting is that md files can just import their own compenents?
      - Why? because importing components in markdown requires a way to resolve those imports which means running "on disk"

**Comment by demenech** [comment]
## Updates 2023-16-03

The replacement is complete.

## Changes:

- MarkdownDB is now a library on its own, see libs/markdowndb
   - Types are now defined on a different file
   - It's now possible to query by _url_path (like a slug) 
   - MarkdownDB now handles the creation of the database
   - MarkdownDB is now capable of not indexing files based on regex patterns (needed because: 1-  we are not validating mdx files yet 2-  next build fails if a certain path provided by getStaticPaths doesn't work 3-  the mdx files in the Excalidraw folder are not valid)
- indexFolder is now run on next.config.mjs
- Removed search script and page
- Rendering pipeline now has basic layout support
- Fixed getters and migrated index.md files for:
  - Docs
  - Blogs
  - Collections
- Migrated content/index.md to pages/index.tsx and content/enterprise/index.md to pages/enterprise.tsx 
- Contentlayer dependencies and configurations are completely removed, we are officially running on MarkdownDB!
- Fixed tests
___
**PR:**  https://github.com/datopian/datahub-next/pull/56

**Comment by demenech** [comment]
@rufuspollock this one is ready to be reviewed.

After that, we can also close https://github.com/datopian/datahub-next/issues/32

**Comment by rufuspollock** [comment]
FIXED. All done - resolved in PR #56
